variable/include/scipp/variable/bucket_variable.tcc

```diff
@@ -76,7 +78,32 @@ auto contiguous_indices(const VariableConstView &parent,
 }
 } // namespace
 
-template <class T> class BucketVariableMaker : public AbstractVariableMaker {
+template <class T> class BinVariableMakerCommon : public AbstractVariableMaker {
+public:
+  bool is_bins() const override { return true; }
+  Variable empty_like(const VariableConstView &prototype,
+                      const std::optional<Dimensions> &shape,
+                      const VariableConstView &sizes) const override {
+    if (shape)
+      throw except::TypeError(
+          "Cannot specify shape in `empty_like` for non-bin prototype, shape "
+          "is given by shape of `sizes`.");
+    const auto [indices, dim, buf] = prototype.constituents<bucket<T>>();
+    Variable sizes_owner;
```

Comment:
Could use
```.cpp
const auto [size_, _] = [&]() {
  if (sizes) {
    return std::tuple{sizes, Variable{}};
  }
  else {
     const auto &[begin, end] = unzip(indices);
     Variable sizes_owner = end - begin;
     return std::tuple{VariableConstView{sizes_owner}, std::move(sizes_owner)};
  }
}();
```
(Is moving the owner safe for the view?)
Could be considered cleaner and might or might not be more efficient.

Reply:
No, the owner may not be moved, since the view stores a pointer to it (rather than the internal array).

Reply:
That sounds pretty fragile. Does it impose any limitations outside of cases like the above?

Reply:
The alternative would be to introduce a shared ownership model for the underlying data buffers, as mentioned recently. This is a larger discussion we will have next year, since it impacts many things. Views turning invalid if variables are moved is unfortunately the least of the problems. We treat views generally as short lived, like iterators.

Reply:
Ok, I'll shut up about for for now, then :-)

Reply:
I think this is maybe a good example of of the wider ownership issues to discuss next year. I can see why this needs be be done as is for the current behaviour, though a comment about `sizes_owner` wouldn't hurt.